File: libs/power-log-parser/src/lib/parsers/cost-changed-parser.ts
import { GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider, GameEventHelper } from '../game-event';
import { Node, NodeType, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class CostChangedParser implements ActionParser {
	readonly ParserName = 'CostChangedParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, facade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = facade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.TagChange &&
			(node.Object as TagChange).Name === (GameTag.COST as number)
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const entity = this.GameState.CurrentEntities.get(tagChange.Entity) ?? null;
		if (entity == null) {
			return null;
		}

		if (entity.GetTag(GameTag.LETTUCE_IS_EQUPIMENT) === 1) {
			return null;
		}

		const cardId = !entity.CardId ? null : entity.CardId;
		const controllerId = entity.GetEffectiveController();
		const abilityOwner = entity.GetTag(GameTag.LETTUCE_ABILITY_OWNER);
		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				'COST_CHANGED',
				GameEventHelper.CreateProvider(
					'COST_CHANGED',
					cardId as any,
					controllerId,
					entity.Id,
					this.StateFacade,
					{
						NewCost: tagChange.Value,
						AbilityOwnerEntityId: abilityOwner,
					},
				),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/CostChangedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class CostChangedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public CostChangedParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                // This is for now only useful to get the speed update in Mercenaries, so we try to restrict the number of events
                // Need it to properly support Lady Prestor + Dredge / discover in your own deck
                //&& ParserState.IsMercenaries()
                && node.Type == typeof(TagChange)
                && (node.Object as TagChange).Name == (int)GameTag.COST;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            FullEntity entity = GameState.CurrentEntities.TryGetValue(tagChange.Entity, out entity) ? entity : null;
            if (entity == null)
            {
                return null;
            }

            // Equipment shouldn't have their cost changed
            if (entity.GetTag(GameTag.LETTUCE_IS_EQUPIMENT) == 1)
            {
                return null;
            }
             

            var cardId = string.IsNullOrEmpty(entity.CardId) ? null : entity.CardId;
            var controllerId = entity.GetEffectiveController();
            var abilityOwner = entity.GetTag(GameTag.LETTUCE_ABILITY_OWNER);
            return new List<GameEventProvider> { GameEventProvider.Create(
                    tagChange.TimeStamp,
                    "COST_CHANGED",
                    GameEvent.CreateProvider(
                        "COST_CHANGED",
                        cardId,
                        controllerId,
                        entity.Id,
                        StateFacade,
                        //null,
                        new {
                            NewCost = tagChange.Value,
                            AbilityOwnerEntityId = abilityOwner,
                        }),
                    true,
                    node) };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
